sdks/java/io/mongodb/src/main/java/org/apache/beam/sdk/io/mongodb/MongoDbGridFSIO.java

```diff
@@ -73,132 +79,136 @@
+      try (BufferedReader reader =
+        new BufferedReader(new InputStreamReader(input.getInputStream()))) {
+        String line = reader.readLine();
+        while (line != null) {
+          result.outputWithTimestamp(line, time);
+          line = reader.readLine();
         }
-
-        @Override
-        public void remove() {
-          throw new UnsupportedOperationException("Remove not supported");
-        }
-      };
+      }
     }
   }
 
   /** Read data from GridFS. */
   public static Read<String> read() {
-    return new Read<String>(new Read.BoundedGridFSSource<String>(null, null, null, null,
-                            StringsParseCallback.INSTANCE, StringUtf8Coder.of()));
+    return new Read<String>(new Read.BoundedGridFSSource(null, null, null, null,
+                            null), StringParser.INSTANCE, Duration.ZERO);
   }
 
   static class Read<T> extends PTransform<PBegin, PCollection<T>> {
     public Read<T> withUri(String uri) {
-      return new Read<T>(new BoundedGridFSSource<T>(uri, options.database,
+      return new Read<T>(new BoundedGridFSSource(uri, options.database,
                                            options.bucket, options.filterJson,
-                                           options.parser, options.coder));
+                                           null), parser, allowedTimestampSkew);
     }
 
     public Read<T> withDatabase(String database) {
-      return new Read<T>(new BoundedGridFSSource<T>(options.uri, database,
+      return new Read<T>(new BoundedGridFSSource(options.uri, database,
                                            options.bucket, options.filterJson,
-                                           options.parser, options.coder));
+                                           null), parser, allowedTimestampSkew);
     }
 
     public Read<T> withBucket(String bucket) {
-      return new Read<T>(new BoundedGridFSSource<T>(options.uri, options.database, bucket,
-          options.filterJson, options.parser, options.coder));
+      return new Read<T>(new BoundedGridFSSource(options.uri, options.database, bucket,
+          options.filterJson, null), parser, allowedTimestampSkew);
     }
 
-    public <X> Read<X> withParsingFn(ParseCallback<X> f) {
-      return new Read<X>(new BoundedGridFSSource<X>(options.uri, options.database,
-          options.bucket, options.filterJson, f, null));
+    public <X> Read<X> withParser(Parser<X> f) {
+      Preconditions.checkNotNull(f, "Parser cannot be null");
+      return new Read<X>(new BoundedGridFSSource(options.uri, options.database,
+          options.bucket, options.filterJson, null), f, allowedTimestampSkew);
     }
-
-    public Read<T> withCoder(Coder<T> coder) {
-      return new Read<T>(new BoundedGridFSSource<T>(options.uri, options.database,
-          options.bucket, options.filterJson, options.parser, coder));
+    public Read<T> allowedTimestampSkew(Duration skew) {
```

Comment:
Hmm, where did this come from and why was it not necessary in the previous version? Is it, logically, something that only the user can specify a proper value for? It could use a javadoc.